### ISSUE TITLE
Failed dependencies on AWS deployment 

### ISSUE THREAD

**Issue by guerrysemones** [miguno/wirbelsturm#36]
Hi,

I'm attempting to use wirbelstrum on AWS. This is my first experience with wirbelsturm and puppet.

I set up an instance as the primary box from which to run wirbelsturm. After getting that set up, I followed all of the AWS directions, and that all seemed to go fine. When I run "vagrant up" on my primary box, it successfully launches all of the instances on EC2. However, each instance (kafka1, nimbus1, supervisor1, etc.) on the deployment gets failed dependencies (listing for kafka1 section  below).

The boxes seem to launch fine, but it appears the dependency failure is complete in that no Storm, Kafka, etc. services are running. Which is probably obvious.

Immediately I get this warning:

```
==> kafka1: Warning! The AWS provider doesn't support any of the Vagrant
==> kafka1: high-level network configurations (`config.vm.network`). They
==> kafka1: will be silently ignored.
```

The first error is:
```
==> kafka1: Error: Execution of '/usr/bin/yum -d 0 -e 0 -y list kafka' returned 1: Error: No matching Packages to list
```

Followed later by: 
```
==> kafka1: Error: Execution of '/usr/bin/yum -d 0 -e 0 -y list supervisor' returned 1: Existing lock /var/run/yum.pid: another copy is running as pid 4906.
==> kafka1: Another app is currently holding the yum lock; waiting for it to exit...
==> kafka1:   The other application is: yum
==> kafka1:     Memory :  49 M RSS (260 MB VSZ)
==> kafka1:     Started: Thu Feb 19 03:30:52 2015 - 00:01 ago
==> kafka1:     State  : Running, pid: 4906
==> kafka1: Error: No matching Packages to list
==> kafka1: Error: /Stage[main]/Supervisor/Package[supervisor]/ensure: change from absent to present failed: Execution of '/usr/bin/yum -d 0 -e 0 -y list supervisor' returned 1: Existing lock /var/run/yum.pid: another copy is running as pid 4906.
```

Any pointers or advice?

Thanks in advance! 

Guerry 

(kafka1 listing below).

```
------------------------------
==> kafka1: Notice: Scope(Class[main]): Deployment environment: 'staging-environment'
==> kafka1: Notice: Scope(Class[main]): I am running within a Vagrant-controlled environment
==> kafka1: Notice: Scope(Class[main]): Disabling firewall...
==> kafka1: Notice: Scope(Class[main]): I have been assigned the role 'kafka_broker'
==> kafka1: Notice: Compiled catalog for kafka1.gateway01.vintel.info in environment production in 1.66 seconds
==> kafka1: Notice: /Stage[main]/Firewall::Linux::Redhat/Service[iptables]/enable: enable changed 'true' to 'false'
==> kafka1: Notice: /Stage[main]/Kafka::Users/Group[kafka]/ensure: created
==> kafka1: Notice: /Stage[main]/Kafka::Users/User[kafka]/ensure: created
==> kafka1: Notice: /Stage[main]/Kafka::Install/File[/var/log/kafka]/ensure: created
==> kafka1: Notice: /Stage[main]/Kafka::Install/Kafka::Install::Create_log_dirs[/app/kafka/log]/Exec[create-kafka-log-directory-/app/kafka/log]/returns: executed successfully
==> kafka1: Error: Execution of '/usr/bin/yum -d 0 -e 0 -y list kafka' returned 1: Error: No matching Packages to list
==> kafka1: Error: /Stage[main]/Kafka::Install/Package[kafka]/ensure: change from absent to present failed: Execution of '/usr/bin/yum -d 0 -e 0 -y list kafka' returned 1: Error: No matching Packages to list
==> kafka1: Notice: /Stage[main]/Kafka::Install/File[/opt/kafka/logs]: Dependency Package[kafka] has failures: true
==> kafka1: Warning: /Stage[main]/Kafka::Install/File[/opt/kafka/logs]: Skipping because of failed dependencies
==> kafka1: Notice: /Stage[main]/Wirbelsturm_common::Install/File[/var/lib/cloud/data/scripts]/ensure: created
==> kafka1: Notice: /Stage[main]/Wirbelsturm_common::Install/Package[jq]/ensure: created
==> kafka1: Notice: /Stage[main]/Kafka::Install/Kafka::Install::Create_log_dirs[/app/kafka/log]/File[kafka-log-directory-/app/kafka/log]/owner: owner changed 'root' to 'kafka'
==> kafka1: Notice: /Stage[main]/Kafka::Install/Kafka::Install::Create_log_dirs[/app/kafka/log]/File[kafka-log-directory-/app/kafka/log]/group: group changed 'root' to 'kafka'
==> kafka1: Notice: /Stage[main]/Kafka::Install/Kafka::Install::Create_log_dirs[/app/kafka/log]/File[kafka-log-directory-/app/kafka/log]/mode: mode changed '0755' to '0750'
==> kafka1: Notice: /Stage[main]/Kafka::Config/File[/opt/kafka/config/log4j.properties]: Dependency Package[kafka] has failures: true
==> kafka1: Warning: /Stage[main]/Kafka::Config/File[/opt/kafka/config/log4j.properties]: Skipping because of failed dependencies
==> kafka1: Notice: /Stage[main]/Kafka::Config/File[/opt/kafka/config/server.properties]: Dependency Package[kafka] has failures: true
==> kafka1: Warning: /Stage[main]/Kafka::Config/File[/opt/kafka/config/server.properties]: Skipping because of failed dependencies
==> kafka1: Error: Execution of '/usr/bin/yum -d 0 -e 0 -y list supervisor' returned 1: Existing lock /var/run/yum.pid: another copy is running as pid 4906.
==> kafka1: Another app is currently holding the yum lock; waiting for it to exit...
==> kafka1:   The other application is: yum
==> kafka1:     Memory :  49 M RSS (260 MB VSZ)
==> kafka1:     Started: Thu Feb 19 03:30:52 2015 - 00:01 ago
==> kafka1:     State  : Running, pid: 4906
==> kafka1: Error: No matching Packages to list
==> kafka1: Error: /Stage[main]/Supervisor/Package[supervisor]/ensure: change from absent to present failed: Execution of '/usr/bin/yum -d 0 -e 0 -y list supervisor' returned 1: Existing lock /var/run/yum.pid: another copy is running as pid 4906.
==> kafka1: Another app is currently holding the yum lock; waiting for it to exit...
==> kafka1:   The other application is: yum
==> kafka1:     Memory :  49 M RSS (260 MB VSZ)
==> kafka1:     Started: Thu Feb 19 03:30:52 2015 - 00:01 ago
==> kafka1:     State  : Running, pid: 4906
==> kafka1: Error: No matching Packages to list
==> kafka1: Notice: /Stage[main]/Supervisor/File[/var/run/supervisor]: Dependency Package[supervisor] has failures: true
==> kafka1: Warning: /Stage[main]/Supervisor/File[/var/run/supervisor]: Skipping because of failed dependencies
==> kafka1: Notice: /Stage[main]/Supervisor/File[/var/log/supervisor]: Dependency Package[supervisor] has failures: true
==> kafka1: Warning: /Stage[main]/Supervisor/File[/var/log/supervisor]: Skipping because of failed dependencies
==> kafka1: Notice: /Stage[main]/Supervisor/File[/etc/logrotate.d/supervisor]: Dependency Package[supervisor] has failures: true
==> kafka1: Warning: /Stage[main]/Supervisor/File[/etc/logrotate.d/supervisor]: Skipping because of failed dependencies
==> kafka1: Notice: /Stage[main]/Supervisor/File[/etc/supervisord.d]: Dependency Package[supervisor] has failures: true
==> kafka1: Warning: /Stage[main]/Supervisor/File[/etc/supervisord.d]: Skipping because of failed dependencies
==> kafka1: Notice: /Stage[main]/Supervisor/File[/etc/supervisord.conf]: Dependency Package[supervisor] has failures: true
==> kafka1: Warning: /Stage[main]/Supervisor/File[/etc/supervisord.conf]: Skipping because of failed dependencies
==> kafka1: Notice: /Stage[main]/Supervisor/Service[supervisord]: Dependency Package[supervisor] has failures: true
==> kafka1: Warning: /Stage[main]/Supervisor/Service[supervisord]: Skipping because of failed dependencies
==> kafka1: Notice: /Stage[main]/Kafka::Service/Exec[restart-kafka-broker]: Dependency Package[kafka] has failures: true
==> kafka1: Notice: /Stage[main]/Kafka::Service/Exec[restart-kafka-broker]: Dependency Package[supervisor] has failures: true
==> kafka1: Warning: /Stage[main]/Kafka::Service/Exec[restart-kafka-broker]: Skipping because of failed dependencies
==> kafka1: Notice: /Stage[main]/Kafka::Service/Supervisor::Service[kafka-broker]/File[/var/log/supervisor/kafka-broker]: Dependency Package[kafka] has failures: true
==> kafka1: Notice: /Stage[main]/Kafka::Service/Supervisor::Service[kafka-broker]/File[/var/log/supervisor/kafka-broker]: Dependency Package[supervisor] has failures: true
==> kafka1: Warning: /Stage[main]/Kafka::Service/Supervisor::Service[kafka-broker]/File[/var/log/supervisor/kafka-broker]: Skipping because of failed dependencies
==> kafka1: Notice: /Stage[main]/Kafka::Service/Supervisor::Service[kafka-broker]/File[/etc/supervisord.d/kafka-broker.conf]: Dependency Package[kafka] has failures: true
==> kafka1: Notice: /Stage[main]/Kafka::Service/Supervisor::Service[kafka-broker]/File[/etc/supervisord.d/kafka-broker.conf]: Dependency Package[supervisor] has failures: true
==> kafka1: Warning: /Stage[main]/Kafka::Service/Supervisor::Service[kafka-broker]/File[/etc/supervisord.d/kafka-broker.conf]: Skipping because of failed dependencies
==> kafka1: Notice: /Stage[main]/Kafka::Service/Supervisor::Service[kafka-broker]/Service[supervisor::kafka-broker]: Dependency Package[kafka] has failures: true
==> kafka1: Notice: /Stage[main]/Kafka::Service/Supervisor::Service[kafka-broker]/Service[supervisor::kafka-broker]: Dependency Package[supervisor] has failures: true
==> kafka1: Warning: /Stage[main]/Kafka::Service/Supervisor::Service[kafka-broker]/Service[supervisor::kafka-broker]: Skipping because of failed dependencies
==> kafka1: Notice: /Stage[main]/Kafka/Anchor[kafka::end]: Dependency Package[kafka] has failures: true
==> kafka1: Notice: /Stage[main]/Kafka/Anchor[kafka::end]: Dependency Package[supervisor] has failures: true
==> kafka1: Warning: /Stage[main]/Kafka/Anchor[kafka::end]: Skipping because of failed dependencies
==> kafka1: Notice: /Stage[main]/Supervisor::Update/Exec[supervisor::update]: Dependency Package[kafka] has failures: true
==> kafka1: Notice: /Stage[main]/Supervisor::Update/Exec[supervisor::update]: Dependency Package[supervisor] has failures: true
==> kafka1: Warning: /Stage[main]/Supervisor::Update/Exec[supervisor::update]: Skipping because of failed dependencies
==> kafka1: Notice: /Stage[main]/Wirbelsturm_common::Install/Package[java-runtime-environment]/ensure: created
==> kafka1: Notice: Finished catalog run in 17.04 seconds
==> kafka1: An error occurred. The error will be shown after all tasks complete.
------------------------------
```

**Comment by miguno** [comment]
Hi Guerry, thanks for the detailed report.

First, you can safely ignore the warning.  It's a standard warning logged by Vagrant because it can't fully control network-related settings in AWS.

The first error seems to indicate that there is a problem with the yum setup on the machines, possibly due to some issues with yum repositories.  If you try to manually run yum on any of the machines, what are the results?  For instance, what happens if you run `yum info kafka` or `yum info storm`?  (You can also try `yum install ...`).

The subsequent "lock" error is also pointing to yum, though whether it is caused by yum or by something (like Puppet) that uses yum is hard to say.

```
==> kafka1: Error: Execution of '/usr/bin/yum -d 0 -e 0 -y list supervisor' returned 1: Existing lock /var/run/yum.pid: another copy is running as pid 4906.
==> kafka1: Another app is currently holding the yum lock; waiting for it to exit...
==> kafka1:   The other application is: yum
```

Is there anything else running on the EC2 instances?  Like another Puppet-based deployment, or a left-over/zombie from a previous Puppet run?

**Comment by guerrysemones** [comment]
Hi Michael,

I'll try running the yum statements on one or more of the servers today. I think I tried the "list" command last night, but it was late. 

These are AMI as produced by the AWS instructions for Wirbelsturm, with nothing added. When I looked into the one box last night, I didn't see anything running other than the typical OS processes.

I'll also look back at the AWS instructions, particularly the part that builds the AMI. Is there anything in particular I should be looking for? I seem to remember that yum is installing your pre-built RPMs, and that the AWS scripts provided with wirbelsturm creates a repo with those RPMs? 

Thanks,

Guerry

**Comment by guerrysemones** [comment]
Hi,

Okay, I ran this on the kafka1 box:

[ec2-user@kafka1 ~]$ /usr/bin/yum -d 0 -e 0 -y list kafka
Error: No matching Packages to list

Is it possible that I missed a step, or that a step didn't go well when I was doing the AWS setup, such that wirbelsturm cannot find the yum repo for the RPMs?

When I looked for kafka on the box, I found the following. 

[ec2-user@kafka1 ~]$ sudo find / -name "kafka*"
/tmp/vagrant-puppet/manifests-846018e2aa141a5eb79a64b4015fc5f3/hieradata/environments/default-environment/hosts/kafka1.yaml
/tmp/vagrant-puppet/manifests-846018e2aa141a5eb79a64b4015fc5f3/hieradata/roles/kafka_broker.yaml
/tmp/vagrant-puppet/modules-a7e4fc58ae655ec8e8f61357773d8a04/kafka
/tmp/vagrant-puppet/modules-a7e4fc58ae655ec8e8f61357773d8a04/kafka/spec/classes/kafka_spec.rb
/home/kafka
/var/spool/mail/kafka
/var/lib/puppet/reports/kafka1.gateway01.vintel.info
/var/log/kafka
/app/kafka

There were no other processes running on the box beside OS stuff. 

Thanks,

Guerry

**Comment by miguno** [comment]
Can you check whether the Wirbelsturm yum repository is configured under `/etc/yum.repos.d/miguno.repo`?

The contents should look like:

```
[ec2-user@kafka1 ~]$ cat /etc/yum.repos.d/miguno.repo
[miguno]
gpgcheck=1
enabled=1
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-miguno
baseurl=https://s3.amazonaws.com/yum.miguno.com/bigdata/redhat/6/x86_64/
name=Michael G. Noll (@miguno)'s public yum repository for Wirbelsturm
priority=20
```

You should also see Puppet log entries such as the following in the deployment logs:

```
==> kafka1: Notice: /Stage[main]/Wirbelsturm_yumrepos::Epel/File[/etc/pki/rpm-gpg/RPM-GPG-KEY-wirbelsturm-epel-6]/ensure: defined content as '{md5}d865e6b948a74cb03bc3401c0b01b785'
==> kafka1: Notice: /Stage[main]/Wirbelsturm_yumrepos::Epel/Yumrepo[wirbelsturm-epel-6]/ensure: created
==> kafka1: Notice: /Stage[main]/Wirbelsturm_yumrepos::Nginx/File[/etc/pki/rpm-gpg/RPM-GPG-KEY-nginx]/ensure: defined content as '{md5}15b05d0af252f8127772ceda07b35b1f'
==> kafka1: Notice: /Stage[main]/Wirbelsturm_yumrepos::Nginx/Yumrepo[nginx]/ensure: created
==> kafka1: Notice: /Stage[main]/Wirbelsturm_yumrepos::Miguno/File[/etc/pki/rpm-gpg/RPM-GPG-KEY-miguno]/ensure: defined content as '{md5}060e8aca6d6a23a4d9fe43b6de8e2cb7'
==> kafka1: Notice: /Stage[main]/Wirbelsturm_yumrepos::Puppetlabs/File[/etc/pki/rpm-gpg/RPM-GPG-KEY-puppetlabs]/mode: mode changed '0644' to '0444'
...snip...
==> kafka1: Notice: /Stage[main]/Wirbelsturm_yumrepos::Miguno/Yumrepo[miguno]/ensure: created
```

The contents of `/etc/yum/repos.d/*` should similar to:

```
[ec2-user@kafka1 ~]$ ls -l /etc/yum.repos.d/
total 56
-rw-r--r-- 1 root root  668 Sep 18 00:50 amzn-main.repo
-rw-r--r-- 1 root root  728 Feb 20 08:43 amzn-main.repo.rpmsave
-rw-r--r-- 1 root root  324 Sep 18 00:50 amzn-nosrc.repo
-rw-r--r-- 1 root root  382 Feb 20 08:43 amzn-nosrc.repo.rpmsave
-rw-r--r-- 1 root root  686 Sep 18 00:50 amzn-preview.repo
-rw-r--r-- 1 root root  746 Feb 20 08:43 amzn-preview.repo.rpmsave
-rw-r--r-- 1 root root  686 Sep 18 00:50 amzn-updates.repo
-rw-r--r-- 1 root root  746 Feb 20 08:43 amzn-updates.repo.rpmsave
-rw-r--r-- 1 root root  957 Mar  1  2013 epel.repo
-rw-r--r-- 1 root root 1056 Mar  1  2013 epel-testing.repo
-rw-r--r-- 1 root root  236 Feb 20 09:18 miguno.repo
-rw-r--r-- 1 root root  188 Feb 20 09:18 nginx.repo
-rw-r--r-- 1 root root 1262 Apr 23  2014 puppetlabs.repo
-rw-r--r-- 1 root root  275 Feb 20 09:18 wirbelsturm-epel-6.repo
```

I redeployed a cluster to AWS a few minutes ago and everything worked fine.  I did not, however, re-create an AMI or ran through the full AWS initial setup.

**Comment by guerrysemones** [comment]
/etc/yum.repos.d/miguno.repo does not exist.

Here's yum.repos.d:

```
[ec2-user@ip-172-31-62-161 ~]$ ls -la /etc/yum.repos.d/
total 32
drwxr-xr-x  2 root root 4096 Jan 27 21:43 .
drwxr-xr-x 71 root root 4096 Feb 14 21:14 ..
-rw-r--r--  1 root root  668 Sep 18 00:50 amzn-main.repo
-rw-r--r--  1 root root  324 Sep 18 00:50 amzn-nosrc.repo
-rw-r--r--  1 root root  686 Sep 18 00:50 amzn-preview.repo
-rw-r--r--  1 root root  686 Sep 18 00:50 amzn-updates.repo
-rw-r--r--  1 root root  957 Mar  1  2013 epel.repo
-rw-r--r--  1 root root 1056 Mar  1  2013 epel-testing.repo
[ec2-user@ip-172-31-62-161 ~]$
```

Did I perhaps miss an error during the Custom AMI Creation steps here? https://github.com/miguno/wirbelsturm/blob/master/docs/AWS.md#custom-ami-creation

I'll rebuild my AMI and see what happens.

Thanks,

Guerry

**Comment by guerrysemones** [comment]
Sorry, the listing above was wrong. I just built a new AMI from the Custom AMI Creation steps in the link I posted above. 

The script ./aws-prepare-image.sh does a number of yum updates, and puppetlab related repo work. I do not see where the wirbelsturm-epel-6.repo, miguno.repo, and nginx.repo's you list are installed from.

Here's my new AMI's listing....
```
[ec2-user@ip-172-31-50-34 ~]$ ls -la /etc/yum.repos.d/
total 52
drwxr-xr-x  2 root root 4096 Feb 21 03:00 .
drwxr-xr-x 73 root root 4096 Feb 21 03:01 ..
-rw-r--r--  1 root root  668 Sep 18 00:50 amzn-main.repo
-rw-r--r--  1 root root  728 Feb 21 02:39 amzn-main.repo.rpmsave
-rw-r--r--  1 root root  324 Sep 18 00:50 amzn-nosrc.repo
-rw-r--r--  1 root root  382 Feb 21 02:39 amzn-nosrc.repo.rpmsave
-rw-r--r--  1 root root  686 Sep 18 00:50 amzn-preview.repo
-rw-r--r--  1 root root  746 Feb 21 02:39 amzn-preview.repo.rpmsave
-rw-r--r--  1 root root  686 Sep 18 00:50 amzn-updates.repo
-rw-r--r--  1 root root  746 Feb 21 02:39 amzn-updates.repo.rpmsave
-rw-r--r--  1 root root  957 Mar  1  2013 epel.repo
-rw-r--r--  1 root root 1056 Mar  1  2013 epel-testing.repo
-rw-r--r--  1 root root 1262 Feb 21 03:00 puppetlabs.repo
```

**Comment by miguno** [comment]
The AMI we create will only contain the bare minimum.  The "real" deployment and provisioning is done by Puppet on top of what the AMI provides.  The miguno.repo etc. are pushes via Puppet, so you won't see them on a vanilla AMI.

Do you find the miguno.repo on your host machine, i.e. somewhere in your Wirbelsturm checkout?

If you don't, then maybe you didn't run the `./bootstrap` command, ot the command failed in some way.  The bootstrap will fetch all required Puppet modules to your host machine and put those files under the `puppet/` subdir, and the files will include miguno.repo.

**Comment by guerrysemones** [comment]
I do not find miguno.repo on the host machine, either in the Wirbelsturm checkout or otherwise.

I reran bootstrap. I observed no errors. I started out with a fresh Wirbelsturm checkout, and reran bootstrap. Same result. I could still not find miguno.repo on the host box. Under the wirbelsturm/puppet directory, there are these:

./modules/wirbelsturm_yumrepos/manifests/miguno.pp
./.tmp/librarian/cache/source/git/ae5558d17e7e814b/manifests/miguno.pp

Sorry, still digging at this. My host box is an Amazon Linux instance.

Thanks,

Guerry
